Initial PowerShelll kernel prototype

TylerLeonhardt opened the pull request:
Original PR: https://github.com/dotnet/try/pull/751 (all feedback addressed)

This handles the 3 basic messages:
* HandleSubmitCode
* HandleRequestCompletion
* HandleCancelCurrentCommand (implemented but waiting until dotnet-interactive supports it better)

This change also includes tests 🙂 

Example [PowerShell notebook](https://gist.github.com/TylerLeonhardt/ff9a20fb53313bf31e6c0107b0fec4e3) that came from this very change.

cc @SteveL-MSFT @daxian-dbw

TylerLeonhardt commented:
Ok I believe I've addressed all feedback from the original PR and cleaned up some copy-paste relics.

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/Modules/Microsoft.DotNet.Interactive.PowerShell/Microsoft.DotNet.Interactive.PowerShell.psd1:
```diff
@@ -0,0 +1,131 @@
+# Generated by: tyleonha
+#
+# Generated on: 1/23/2020
+#
```
Replace this comment section with the following?
```
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
//
```

jonsequitur replied:
The copyright and license we're using elsewhere is:

```
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```

TylerLeonhardt replied:
updated 👍 Not sure what you want the AuthorName and CompanyName to be in the psd1 so I left it at Microsoft for now. Let me know if you prefer something else.

SteveL-MSFT replied:
Should probably be `.NET Foundation` instead of `Microsoft`

TylerLeonhardt commented:
Not really sure what to do about the CI failure:
> .packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20073.17\tools\Sign.proj(81,5): error SIGN001: Signing 3rd party library 'D:\a\1\s\artifacts\tmp\Release\ContainerSigning\200\tools/netcoreapp3.1/any/Markdig.Signed.dll' with Microsoft certificate 'Microsoft400'. The library is considered 3rd party library due to its copyright: 'Alexandre Mutel'.

brettfo commented:
@TylerLeonhardt For the `Markdig.Signed.dll` issue, you'll have to add an entry to [this file](https://github.com/dotnet/interactive/blob/master/eng/Signing.props#L32).

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,272 @@
+            }
+            else
+            {
+                context.Publish(new IncompleteCodeSubmissionReceived(submitCode));
```
Does `IncompleteCodeSubmissionReceived` literally means the code is incomplete? `IsCompleteSubmission` returns `true` as long as the script contains parsing error, but it could be complete. My concern is about what message will the user see in that case.

jonsequitur replied:
Yes, this is the intent. This is important for languages that allow multi-line statements. Something like this (in C#) would be valid in a REPL:

```csharp
var x =
```

but not in a notebook, where we don't want statements to break across cells because it makes the current state of the notebook hard to understand.

daxian-dbw replied:
Thanks @jonsequitur for the clarification!
Then `IsCompleteSubmission` should explicitly check for `System.Management.Automation.IncompleteParseException`.

TylerLeonhardt replied:
@daxian-dbw does `Parser.ParseInput` throw an `IncompleteParseException` or is that exception wrapped in a `ParseError`?

TylerLeonhardt replied:
I've followed what PowerShell does (checking the first ParseError for if the input is Incomplete

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,272 @@
+            }
+        
+            // Wait until we have access to the runspace since we only can run one cell at a time.
+            await _runspaceSemaphore.WaitAsync();
```
Is it possible that multiple requests coming in parallel? A Jupyter client will spins up a kernel, but can a kernel be shared by multiple clients?

TylerLeonhardt replied:
Yeah you can trigger multiple cells at the same time so this protects the runspace from a second invocation trying to take the runspace.

jonsequitur replied:
The command pipeline should prevent parallel execution. While the Jupyter client lets you send a new submission  while a previous one is executing, the kernel forces them to run serially.

daxian-dbw replied:
Great to know that (I guess that's done by the [Aggregattion here](https://github.com/dotnet/interactive/blob/master/Microsoft.DotNet.Interactive/KernelCommandPipeline.cs#L59-L63)).
Then we probably can remove the semaphore.

TylerLeonhardt replied:
What about the completion message? Would that also wait for the execution of code to finish before handling the completion request?

jonsequitur replied:
The trampoline here is intended to enforce serial execution regardless of the kind of command sent to the kernel: https://github.com/dotnet/interactive/blob/a49941b40730cebc65dd73af6c65401e0a9fc06b/Microsoft.DotNet.Interactive/KernelBase.cs#L192-L202

If you see any commands executing in parallel, consider it a bug.

TylerLeonhardt replied:
I've tested this and they look like they're handled serially as well.

TylerLeonhardt replied:
I've removed the semaphore 👍

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,272 @@
+                        "dotnet",
+                        "cs"));
+
+            await new KernelExtensionAssemblyLoader().LoadFromAssembliesInDirectory(
```
Why not have a `_extensionLoader` and initialize it in the constructor like the CSharp kernel?

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,272 @@
+            }
+        }
+
+        public async Task LoadExtensionsFromDirectory(
```
A comment saying that we are loading c# extensions would be useful.

TylerLeonhardt replied:
done

colombod replied:
this kernel should not be extensible unless you clearly want that. Here you are just loading extensions that have been developed for csharp kernel.

We should discuss this part better but you are note required to be implement ```IExtensibleKernel```.

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell.Tests/Microsoft.DotNet.Interactive.PowerShell.Tests.csproj:
```diff
@@ -0,0 +1,19 @@
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <Deterministic Condition="'$(NCrunch)' == '1'">false</Deterministic>
```
Please add `<LangVersion>latest</LangVersion>` here.

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/Microsoft.DotNet.Interactive.PowerShell.csproj:
```diff
@@ -0,0 +1,27 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Please add `<LangVersion>latest</LangVersion>` here.

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,272 @@
+
+            await new KernelExtensionAssemblyLoader().LoadFromAssembliesInDirectory(
+                extensionsDirectory,
+                context.HandlingKernel,
```
Can this be replaced with `this`?

TylerLeonhardt replied:
yes! Replaced.

TylerLeonhardt commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,272 @@
+            {
+                context.Fail(null, "Command cancelled");
+                return;
+            }
```
That's possible. I'll probably need to borrow code from C# subkernel. Let's defer that for this PR if you don't mind.

daxian-dbw replied:
No, I'm don't mind at all. This work can be done later.

TylerLeonhardt commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,272 @@
+
+            Environment.SetEnvironmentVariable("PSModulePath",
+                psModulePath + Path.PathSeparator + Path.Join(
+                    Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
```
fixed

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/Commands/TracePipelineObjectCommand.cs:
```diff
@@ -20,7 +20,7 @@ public sealed class TracePipelineObjectCommand : PSCmdlet
         /// The object from pipeline.
         /// </summary>
-        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true)]
+        [Parameter(Position = 0, ValueFromPipeline = true)]
```
What's wrong with having the `Mandatory` here?

TylerLeonhardt replied:
If I have a cell that contains just `$thisVarNotDefined` an error is written out saying `trace-pipelineobject 's InputObject parameter is mandatory and can't handle null`

daxian-dbw replied:
I see. Then we should add `[AllowNull]` instead of removing `Mandatory = true`.

TylerLeonhardt replied:
added!

TylerLeonhardt commented:
rebased and added better support for parse errors:

![image](https://user-images.githubusercontent.com/2644648/73319466-4fcd4d80-41f1-11ea-9476-9f54d0497549.png)

daxian-dbw commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -108,16 +108,25 @@ public PowerShellKernel()
+                    context.Publish(new CommandFailed(null, submitCode, message: parseError.ToString()));
+                }
+                return;
+            }        
```
There are tailing spaces in this line.

TylerLeonhardt replied:
fixed!

TylerLeonhardt commented:
Can someone help me understand the CI failure?

```
.packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20077.7\tools\Publish.proj(283,5): error : Invalid ELF BuildID '<null>' for tools/netcoreapp3.1/any/runtimes/linux-x64/native/libmi.so
```

jonsequitur commented:
@brettfo  Does this build failure ring any bells?

colombod commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellExtensions.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace Microsoft.DotNet.Interactive.PowerShell
+{
+    using System.Management.Automation;
```
why nested namespace?

TylerLeonhardt replied:
The nested namespace is because PowerShell is both a part of a namespace, and a type.

colombod commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,274 @@
+            {
+                foreach (ParseError parseError in parseErrors)
+                {
+                    context.Publish(new CommandFailed(null, submitCode, message: parseError.ToString()));
```
this should  be 
```context.Fail(string.Join("\n", parseErrors.Select(pe => pe.ToString()));```

a command cannot Fail more than once.

TylerLeonhardt replied:
changed.

brettfo commented:
@jonsequitur The build failure is due to dotnet/symstore#81.  I've sent email internally to figure out what the workaround is and how to apply it here.  I'll reply once I have something actionable.

jonsequitur commented on Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs:
```diff
@@ -886,5 +887,35 @@ public async Task PowerShell_streams_handled_in_correct_order()
                 e => e.Should().BeOfType<CommandHandled>());
         }
+
+        [Fact(Timeout = 45000)]
```
You don't need to add the timeouts.

TylerLeonhardt replied:
All the other tests in this file have timeouts for similar tests. Should remove those as well?

jonsequitur replied:
If you want to, feel free. If not, we'll do it. These were an artifact of an investigation into timeout and memory issues that were plaguing us but have been resolved.

TylerLeonhardt replied:
I removed them just from my added tests.

brettfo commented:
@TylerLeonhardt Can you try merging/cherry-picking/manually-applying the head commit in the `snupkg` branch?  It may help with the `Invalid ELF BuildID '<null>'` issue.

TylerLeonhardt commented:
@brettfo doesn't look like that did the trick :(

brettfo commented:
I just added a commit that manually strips the offending files from the symbols packages.  Build worked locally, now we'll see what happens in the PR.

brettfo commented:
The recent MSBuild hackery gets us past the symbol uploading and the Linux leg now passes.

I'll pass this back to @TylerLeonhardt now to look into the test failure(s) in Windows and add any extra features/tests you may want.

colombod commented on Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs:
```diff
@@ -271,6 +277,10 @@ public async Task it_returns_diagnostics(Language language)
             {
                 Language.FSharp => "input.fsx (1,1)-(1,7) typecheck error The value or constructor 'aaaadd' is not defined.",
                 Language.CSharp => "(1,1): error CS0103: The name 'aaaadd' does not exist in the current context",
+                Language.PowerShell => @"At line:1 char:5
```
This is a verbatim string, new lines will be captured from the OS where the code was prepared and not from the current machine. This could lead to flaky tests. if you are testing multilines maybe better us string join with the OS NewLine character or find other ways.

TylerLeonhardt replied:
This test discovered a bug in PowerShell - a hardcoded `\n` https://github.com/PowerShell/PowerShell/blob/ed1f6e3144a2f6fa78e69582870e29ae898e1e78/src/System.Management.Automation/engine/parser/Parser.cs#L8062

I'll be sending a PR to PowerShell to fix that but as a workaround, I'm replacing `\r\n` with `\n` which will work generically.

TylerLeonhardt replied:
PR to fix PowerShell: https://github.com/PowerShell/PowerShell/pull/11746

TylerLeonhardt commented:
Green CI 🤩

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs:
```diff
@@ -0,0 +1,256 @@
+            _runspace.Open();
+            _pwsh = PowerShell.Create(_runspace);
+            _cancellationSource = new CancellationTokenSource();
+            _extensionLoader = new KernelExtensionAssemblyLoader();
```
This isn't need for now, right?

TylerLeonhardt replied:
good catch - removed.

jonsequitur commented on Microsoft.DotNet.Interactive.PowerShell/StreamHandler.cs:
```diff
@@ -0,0 +1,199 @@
+        private static readonly object _pwshLock = new object();
+        KernelInvocationContext _context;
+        IKernelCommand _command;
+        ConcurrentDictionary<int, DisplayedValue> _progresses;
```
Please specify `private` and, where possible, `readonly`.

TylerLeonhardt replied:
fixed.

TylerLeonhardt commented:
🎉🎉🎉🎉

brettfo commented:
@TylerLeonhardt The full-signed build with the PowerShell kernel failed due to signing issues (link [here](https://dev.azure.com/dnceng/internal/_build/results?buildId=503911&view=logs&j=b11b921d-8982-5bb3-754b-b114d42fd804&t=cdcedd1f-8008-523f-9da1-cc35fbfef9a3)).  Do you know who owns the publishing of the `Microsoft.Management.Infrastructure` package?  Looking at the NuGet packages, there are different public keys used for Unix vs. Windows binaries, and not everything is consistent.  I'm not aware of a workaround on our end that could avoid checking those assemblies, but I'll keep looking.

I tried using the `3PartySHA2` key in `eng/Signing.props`, but that doesn't pass build, while the `None` key currently specified passes build but not official signing validation.

My guess is that this package (and it's Unix and Windows sub-packages) will have to be re-published with consistent signing/keys.

TylerLeonhardt commented:
@adityaPatwardhan @travisez13 @SteveL-MSFT can one of you guys offer some advice here? I'm not familiar with this package.

brettfo commented:
After speaking with some people, I found a way to exclude just that one file from the signing checks due to the different signatures.  See PR #110.
